### ISSUE TITLE
Animate window move/resize/swap/warp 

### ISSUE THREAD

**Issue by dominiklohmann** [koekeishiya/yabai#148]
### Discussion / Feature request

This is a long shot, and probably far off in the future. This has been previously discussed in https://github.com/koekeishiya/chunkwm/issues/144.

I've dug up some stuff regarding this.

- How Chromium resizes and moves windows using `CGS*` functions  https://github.com/ChromiumWebApps/chromium/blob/c7361d39be8abd1574e6ce8957c8dbddd4c6ccf7/chrome/browser/ui/cocoa/constrained_window/constrained_window_animation.mm
- A (very old) blog post explaining how `CGSSetWindowWarp` works http://kevin.sb.org/2006/07/23/cgssetwindowwarp-explained/
- An working example for using `CGSSetWindowWarp` and `CGSSetWindowTranslation` for custom animations https://github.com/sailesha/CGSSetWindowWarp-Sample/blob/master/windowAnimation.mm

Obviously all the above examples require access to the NSWindow instance for async animations, which is inaccessible from the Dock process—which is the only process yabai injects code into.

Simply calling `SLSSetWindowTransformation` with a `CGAffineTransform` (e..g `CGAffineTransformMakeTranslation(100, 100)` to move a window 100 right and 100 down) will move the window, but this is not animated. Calling this repeatedly will likely cause lag.

And now here's the long shot that I'd love to see investigated in the future: There are additional functions called `SLSTransactionCreate` and `SLSTransactionCommit`, with variants of the transform and warp functions named `SLSTransactionWindowTransform` and so on. Assuming this works similar to `CATransaction` on iOS, this could then be used to animate multiple windows with a duration and at the same time.

These `SLSTransaction*` functions are available (on 10.15 dev beta 4 currently):

<details>
<summary><code>nm /System/Library/PrivateFrameworks/SkyLight.framework/SkyLight | grep SLSTransaction</code></summary>

```
0000000000200d1e T _SLSTransactionAddWindowToSpace
0000000000200f3a T _SLSTransactionAddWindowToSpaceAndRemoveFromSpaces
0000000000201f5c T _SLSTransactionBindSurface
0000000000200129 T _SLSTransactionClearMenuBarSystemOverrideAlphas
00000000001ffce4 T _SLSTransactionClearWindowLockedBounds
00000000001ff834 T _SLSTransactionClearWindowSystemLevel
0000000000202cb6 T _SLSTransactionCommit
0000000000202ee0 T _SLSTransactionCommitCoalescing
00000000001fd88b T _SLSTransactionCreate
0000000000201a22 T _SLSTransactionDeferWindowMoveEvents
00000000002013c1 T _SLSTransactionDestroySpace
00000000001fd788 T _SLSTransactionGetTypeID
000000000031b9d0 b _SLSTransactionGetTypeID.once
000000000031b9d8 b _SLSTransactionGetTypeID.typeID
000000000020109f T _SLSTransactionHideSpace
00000000001ffd7d T _SLSTransactionMoveWindowWithGroup
0000000000201457 T _SLSTransactionMoveWindowsToManagedSpace
000000000020205f T _SLSTransactionOrderSurface
00000000001fe097 T _SLSTransactionOrderWindow
00000000001fe196 T _SLSTransactionOrderWindowGroup
0000000000201c0a T _SLSTransactionOverrideAppSleepNotifications
0000000000201eb2 T _SLSTransactionPostBroadcastNotification
0000000000202dd4 t _SLSTransactionPostCommit
0000000000201d2a T _SLSTransactionPostNotificationToConnection
0000000000200dd1 T _SLSTransactionRemoveWindowFromSpace
0000000000200e84 T _SLSTransactionRemoveWindowFromSpaces
000000000020132b T _SLSTransactionResetSpaceMenuBar
00000000001fff5c T _SLSTransactionResetWindow
00000000001fef10 T _SLSTransactionResetWindowSubLevel
0000000000202922 T _SLSTransactionSetBackdropChameleonContribution
000000000020283c T _SLSTransactionSetChameleonUpdatesEnabled
0000000000201ec6 T _SLSTransactionSetClientAdvisory
0000000000201950 T _SLSTransactionSetEventCapture
0000000000201654 T _SLSTransactionSetManagedDisplayCurrentSpace
0000000000201875 T _SLSTransactionSetManagedDisplayIsAnimating
00000000002001a9 T _SLSTransactionSetMenuBarBounds
00000000001ffff5 T _SLSTransactionSetMenuBarSystemOverrideAlpha
0000000000200a84 T _SLSTransactionSetSpaceAbsoluteLevel
0000000000200bea T _SLSTransactionSetSpaceAlpha
0000000000200b37 T _SLSTransactionSetSpaceOrderingWeight
0000000000200957 T _SLSTransactionSetSpaceShape
000000000020054f T _SLSTransactionSetSpaceTransform
0000000000202175 T _SLSTransactionSetSurfaceBounds
0000000000202563 T _SLSTransactionSetSurfaceOpacity
0000000000202666 T _SLSTransactionSetSurfaceResolution
00000000001fe295 T _SLSTransactionSetWindowAlpha
00000000001fe3c6 T _SLSTransactionSetWindowBrightness
00000000001ff8cd T _SLSTransactionSetWindowGlobalClipShape
00000000001fe4f7 T _SLSTransactionSetWindowLevel
00000000001ff9fd T _SLSTransactionSetWindowLockedBounds
00000000001fe66b T _SLSTransactionSetWindowOpaqueShape
00000000001fe79b T _SLSTransactionSetWindowProperty
00000000001fefa9 T _SLSTransactionSetWindowReleasesBackingOnOrderOut
00000000001feace T _SLSTransactionSetWindowResolution
00000000001fed2a T _SLSTransactionSetWindowShape
00000000001fee5a T _SLSTransactionSetWindowSubLevel
00000000001ff08f T _SLSTransactionSetWindowSystemAlpha
00000000001ff77e T _SLSTransactionSetWindowSystemLevel
00000000001ff1c0 T _SLSTransactionSetWindowTransform
00000000001ff5e9 T _SLSTransactionSetWindowWarp
0000000000201009 T _SLSTransactionShowSpace
000000000020158b T _SLSTransactionSpaceTileMoveToSpaceAtIndex
0000000000201af2 T _SLSTransactionUpdateRegion
000000000020120c T _SLSTransactionWillSwitchSpaces
00000000001fe615 t __SLSTransactionCommitAction
00000000001fdfe4 t __SLSTransactionDecode_CFString
0000000000052ef0 t __SLSTransactionDecode_CGSSpaceIDArray
00000000001fdc6b t __SLSTransactionDecode_packed64
00000000001fdc9e t __SLSTransactionDecode_packed64_tag
0000000000201135 t __SLSTransactionEncode_CFArray
00000000001fdd41 t __SLSTransactionEncode_CFString
00000000002016fd t __SLSTransactionEncode_UUIDString
00000000001fdba3 t __SLSTransactionEncode_packed64
00000000001fd994 t __SLSTransactionEncode_resize
00000000001fd7cd t __SLSTransactionFinalize
00000000001fd925 t __SLSTransactionPopWriteStream
0000000000201d3a t __SLSTransactionPostNotification
0000000000202a53 t __SLSTransactionPushWriteStream
0000000000202bf6 t __SLSTransactionResetWriteStream
000000000031b9e8 b __SLSTransactionWriteStreamKey.key
000000000031b9e0 b __SLSTransactionWriteStreamKey.once
000000000020300f t ___SLSTransactionCommitCoalescing_block_invoke
00000000001fd7b4 t ___SLSTransactionGetTypeID_block_invoke
0000000000201507 t ___SLSTransactionMoveWindowsToManagedSpace_block_invoke
0000000000202ed8 t ___SLSTransactionPerformCommitAction
0000000000202520 t ___SLSTransactionSetSurfaceBounds_block_invoke
0000000000202808 t ___SLSTransactionSetSurfaceResolution_block_invoke
00000000001fe5ef t ___SLSTransactionSetWindowLevel_block_invoke
00000000001feca3 t ___SLSTransactionSetWindowResolution_block_invoke
00000000002012a5 t ___SLSTransactionWillSwitchSpaces_block_invoke
00000000001fdad3 t ____SLSTransactionWriteStreamKey_block_invoke
```

</details>

Why? Look at this example from i3. Looks nice, doesn't it? 

![28225344-5db44b06-68a0-11e7-93ac-7da7919a6773](https://user-images.githubusercontent.com/4488655/61476477-74688300-a98d-11e9-9651-fbb4a6bfa313.gif)

**Comment by capaldo** [comment]
If not injected how would you go about doing this?

**Comment by koekeishiya** [comment]
I'm not particularly interested in doing this, but would not be against a PR where this is optional.

**Comment by danielfalbo** [comment]
Another animation idea:
it would be really cool to animate the opening of the terminal in a sci-fi style, kinda like [edex-ui](https://github.com/GitSquared/edex-ui) 
![edex](https://user-images.githubusercontent.com/39460524/101246179-f50a5980-3711-11eb-9be3-ce844eb5b175.gif).
Of course it would be cool for other windows too but especially for the terminal because it's usually really flexible in terms of size and is the app that feels the most like a sci-fi thing 🤪.

Alacritty, for example, is completely resizable.
![in](https://user-images.githubusercontent.com/39460524/101246439-a067de00-3713-11eb-9c2d-0a15c104a436.gif)
In the gif I put an Alacritty window in a floating space, and, after trying to make it as small and centered as possible, I first do `yabai -m window --resize left:-20:0` and `yabai -m window --resize right:20:0` multiple times until the window horizontally fits the screen, and then the same thing vertically with `yabai -m window --resize bottom:0:20` and `yabai -m window --resize top:0:-20`

**Comment by i-pwl** [comment]
Any update on this?

**Comment by albert-ying** [comment]
Just saw this on reddit, it looks so smooth

https://www.reddit.com/r/unixporn/comments/vtpwhc/hyprland_dynamic_wallpapers_and_settings_with_a/if9iaoy/?context=3

**Comment by koekeishiya** [comment]
Just to note that the mentioned `SLSTransaction` function family does not work in Monterey 12.5.

**Comment by itaysharir** [comment]
Did anyone manage to implement it and has a compiled version of it?

**Comment by i-pwl** [comment]
im wondering of this too


>

**Comment by FelixKratz** [comment]
https://user-images.githubusercontent.com/22680421/189494638-f8a63c68-297b-41d0-9805-c5cf0f02b33a.mp4

I figured it is not a big problem to call `SLSSetWindowTransformation` frequently to create the intermediate frames. I have hardcoded there to be 100 frames in any resize/move animation.

**Comment by i-pwl** [comment]
brilliant.


> anim.mp4 
> 
> I figured it is not a big problem to call `SLSSetWindowTransformation` frequently to create the intermediate frames. I have hardcoded there to be 100 frames in any resize/move animation.

brilliant.

**Comment by koekeishiya** [comment]
@FelixKratz 

Did you only apply the transform in that POC, or does it also synchronise said transform with the desired window frame change? This is important because macOS itself applies transforms to windows in various context, and that will undo the operation.

**Comment by FelixKratz** [comment]
> @FelixKratz
> 
> Did you only apply the transform in that POC, or does it also synchronise said transform with the desired window frame change? This is important because macOS itself applies transforms to windows in various contexts, and that will undo the operation.

It is not polished at all and very hacked together, this is all I did:
https://github.com/FelixKratz/yabai/commit/2ba29b73df0c9e71cd3d34bf0c97ef2e59cfa7bf
I am only setting the transform for as long as I am animating and then I do the actual proper move/resize. These animations block yabai, which is why I have made them so short.

**Comment by koekeishiya** [comment]
I'll probably take a shot at this for the next release, as I'm working on smoothing out the experience. I have a fairly high bar for what I consider acceptable though, so not sure if it is feasible to ship in the end. A dealbreaker for me is the ability to animate *all* windows that are modified in the same operation simultaneously, so that they start/end at the same time.

**Comment by FelixKratz** [comment]
> I'll probably take a shot at this for the next release, as I'm working on smoothing out the experience. I have a fairly high bar for what I consider acceptable though, so not sure if it is feasible to ship in the end. A dealbreaker for me is the ability to animate _all_ windows that are modified in the same operation simultaneously, so that they start/end at the same time.

Great to hear! Animating the windows in parallel would be very nice, but quite a bit of restructuring of the code base I believe. It would probably also be much better to animate from the SA directly. I only chose to not do that because then the SA would have to notify yabai when the animation is finished and unlock some kind of mutex that the SA and yabai share and that would go too far for a simple POC. This would probably also benefit greatly from your new ipc approach via shmem + mutex. When those animations are done "proper" (parallel and less flickering) it will feel very nice I am sure.

What I stumbled over while implementing the POC was that once the affine transform is reset to the original transform before moving and resizing the window via AX there would sometimes be a noticeable flickering and I didn't find a way to resize the window through SLS functions, i.e. SLSSetWindowShape does not seem to be respected when set via SA. So I was stuck with the slow resize via AX.

**Comment by koekeishiya** [comment]
Pushed a proof of concept myself: https://github.com/koekeishiya/yabai/tree/animate https://github.com/koekeishiya/yabai/commit/269d29d2ddbdd22c4024a0151f461a3dd9def616

Add `yabai -m config window_animation_duration 0.10` to enable animations; adjust the value as you like.
This does not work with window borders yet, so keep borders disabled. This still only animates a single window at a time.

> What I stumbled over while implementing the POC was that once the affine transform is reset to the original transform before moving and resizing the window via AX there would sometimes be a noticeable flickering and I didn't find a way to resize the window through SLS functions, i.e. SLSSetWindowShape does not seem to be respected when set via SA. So I was stuck with the slow resize via AX.

I don't think it is possible to get around this; I have attempted to do so in my version, but no guarantees that it is a universal solution.

**Comment by koekeishiya** [comment]
Transactional:

![animation_sample_mid](https://user-images.githubusercontent.com/6175959/189678090-0f7fc46f-ef2a-470c-9255-1f09e49619c4.gif)

The flicker is caused by the issue mentioned above; when resetting the transform at the end of the animation to apply the change in size using the AX API. Hopefully there will be a workaround to that...

**Comment by koekeishiya** [comment]
Resizing using the AX API is just too  slow. Need to figure out some other way to resize a window.. When resizing a window using the native handle at the edges, the operation appears to be rather snappy. I wonder if we could somehow hook into that.

**Comment by koekeishiya** [comment]
I had an idea that it would maybe be ok to animate the actual windows as is done, but taking a screen capture of the space as the final frame; displaying this screen capture as a topmost window and destroy it (probably some kind of blurred fade) while removing the transforms and applying the resize animation.

Thanks to your comment now I recorded my screen while doing the enter/exit fullscreen application, and it actually looks like Apple themselves are doing something like this, surprisingly enough.

**Comment by FelixKratz** [comment]
I tried my idea by misusing the border to temporarily draw the capture of the window. The scripting addition is not needed for the animation at all and it looks very smooth. https://github.com/FelixKratz/yabai/commit/10793056149a14783662e7a7d6b65ce257b5fef7

**Comment by koekeishiya** [comment]
Are you able to animate the windows simultaneously using that approach though?

**Comment by FelixKratz** [comment]
> Are you able to animate the windows simultaneously using that approach though?
> 
> Edit:
> 
> I guess you could iterate through all windows, grab a capture of the window, and set the opacity to 0. Set the size and position using the AX API.
> 
> Animate all the captures.
> 
> Iterate through the list of windows, setting the opacity back to the previous value.

In my current implementation it is single window, but if I pass a window list to the animation function (same as you have done) it will be possible to animate simultaneously. Maybe I will try to rebase to your POC and implement this method there, but I think the digest of this is that we can definitely circumvent the flickering. Maybe it would be good to use the faster SLSHW functions for window capture instead of what I am using currently.

**Comment by koekeishiya** [comment]
It works pretty well, except that the window order is not the same as it is pre-screenshot, and sometimes the screenshot colours look different from the window itself. e.g Safari is darker than it is in the screenshot.

**Comment by koekeishiya** [comment]
This is probably as good as it is going to get:
![animation](https://user-images.githubusercontent.com/6175959/190002247-aefb15f8-3d5f-4667-9b53-041718789466.gif)


There are some artifacts in the gif due to compression and whatever, it looks much better in reality.

Slight issues with border integration still, but the animation system is pretty much done.

**Comment by FelixKratz** [comment]
I have rebased onto your POC and played with it for a bit. I have made some slight tweaks that get rid of the border flicker and reduce the overhead if borders are activated: https://github.com/FelixKratz/yabai/commit/23ba5bd7085b18797458eb11d507f9ca38b36371

**Comment by koekeishiya** [comment]
@FelixKratz 

Does that solve the border misalignment when using the mouse to move/resize windows (especially noticable when doing a warp operation using the mouse) ? I haven't looked to closely at the border stuff yet.

Does anyone have a set of functions that look nice for animating various kinds of window movement. Might be nice to allow the ability to select between a variety of function, and possibly select an animation based on which action is happening.

**Comment by FelixKratz** [comment]
> @FelixKratz
> 
> Does that solve the border misalignment when using the mouse to move/resize windows (especially noticable when doing a warp operation using the mouse) ? I haven't looked to closely at the border stuff yet.

No, I did not notice it until now since I rarely use my mouse for those actions. But the animation system so far is very nice and helps to indicate direction of movement, especially for other people viewing the screen.

**Comment by FelixKratz** [comment]
> Does anyone have a set of functions that look nice for animating various kinds of window movement. Might be nice to allow the ability to select between a variety of functions, and possibly select an animation based on which action is happening.

* $f(x) = x$
* $f(x) = x^2$
* $f(x) = \sin(x \pi / 2)$
* $f(x) = 1 - \cos(x \pi / 2)$
* $f(x) = x \exp(x - 1)$
* $f(x) = c \tanh(2 \text{artanh}(\frac{1}{2c})(x - 1/2)) + 1/2$, for $c \in (1/2, \infty)$

all of those functions enforce $f(0) = 0$ and $f(1) = 1$.

I am not sure if they are "nice" but they are smooth and infinitely differentiable on the domain. I especially like the last one for $c \sim 0.52$

**Comment by Liquidmantis** [comment]
I don't have anything to add to the discussion, so apologies for the noise, but I just wanted to say thanks to you two for all the effort on this.  It looks amazing and I can't wait.

**Comment by koekeishiya** [comment]
I think I found a way for this entire thing to work without having to disable SIP.

**Comment by FelixKratz** [comment]
I wonder if it would be possible to make these animations not block yabai, similar to how I handle animations in SketchyBar:
https://github.com/FelixKratz/SketchyBar/blob/2008f71c2162c8c4d16f007ebc9f798d8457ba15/src/animation.c#L157-L177
The implication of this would be that something like:
```bash
yabai -m window --toggle zoom-fullscreen \
&& yabai -m query --windows --window | jq '."has-fullscreen-zoom"'
```
would need to return `true` when the animation has begun and not only after it has ended.

**Comment by koekeishiya** [comment]
It is possible to make them non-blocking, yes. However, then we need to handle the case of the user performing some action on a window that is being animated and correctly cancel the animation for that specific window or something, similar to what I did for the window opacity fading system in #1406 

In addition, the queued events will then also be processed while a window is animating, e.g we will receive and process a window resize event for the windows that are animated, while they are being animated, causing their cached properties to update (and it will also trigger a border update).

I'm not sure there is too much to gain by making them run asynchronously.

**Comment by FelixKratz** [comment]
> I'm not sure there is too much to gain by making them run asynchronously.

It could make some things feel a bit better, take for example:
```bash
ctrl + lalt - j    : yabai -m window --resize right:-100:0 || yabai -m window --resize left:-100:0
ctrl + lalt - k    : yabai -m window --resize bottom:0:100 || yabai -m window --resize top:0:100
ctrl + lalt - l    : yabai -m window --resize bottom:0:-100 || yabai -m window --resize top:0:-100
ctrl + lalt - 0x29 : yabai -m window --resize right:100:0 || yabai -m window --resize left:100:0
```
If I hold any of those key combos for a while I am in for a bad time, depending on my key repeat frequency this will create a resize queue several seconds long.

**Comment by koekeishiya** [comment]
> If I hold any of those key combos for a while I am in for a bad time, depending on my key repeat frequency this will create a resize queue several seconds long.

What is the expected behaviour though, cancelling the previous animation or simply queue up the next one? 

I think both scenarios will look pretty janky. I guess it could look smooth by cancelling the current animation and have the new animation start interpolating from the position of the cancelled animation and not the final position (the AX Position of the frame).

**Comment by FelixKratz** [comment]
> What is the expected behaviour though, cancelling the current animation or simply queue up the next one?

At least for me the animation is a way to indicate the spacial directionality of a state change. The state change
happens as soon as I request it, i.e. the state has already changed when the animation is started.
Thus the expected behavior for me is to cancel the current animation and animate to the new state without stopping at the outdated states, because they have been invalidated by my new request.

Of course this is a matter of taste, the way it is currently the animation needs to finish for the state to change, but this leads to the feeling of not being responsive e.g. in the above scenario.

**Comment by koekeishiya** [comment]
I agree that makes sense. I think that would require changes to allow the following which I mentioned above:

> I guess it could look smooth by cancelling the current animation and have the new animation start interpolating from the position of the cancelled animation.

**Comment by koekeishiya** [comment]
async animations poc:

![async_animate](https://user-images.githubusercontent.com/6175959/190739365-6433c745-fee2-41d8-8742-b2d72968bc73.gif)

**Comment by i-pwl** [comment]
can window open and close animations be added?

**Comment by koekeishiya** [comment]
Why is it so hard to make a decent gif..

I think this looks better maybe. Changed the easing function and removed weird flickering that was present in the previous version:

![animation](https://user-images.githubusercontent.com/6175959/190860937-aa845ab5-bd09-4796-848d-6e732f886a45.gif)

**Comment by FelixKratz** [comment]
Wow, it works *really* well and it is exactly how I would expect an animation system to work! Thank you for making this, very nice work!

**Comment by koekeishiya** [comment]
Squashed all changes so far into: https://github.com/koekeishiya/yabai/commit/f9e6e8f42aff7a5c1679abed129e7ab3aa20146f

**Comment by koekeishiya** [comment]
> can window open and close animations be added?

It might be possible to create a window opening animation; depends on how much delay there is between the window spawning and yabai getting notified, I guess this would likely be implemented as a fade+scale transition. 

Animating on window close is pretty much impossible. The window is already inaccessible from system APIs by the time we are notified that a window has been destroyed.

**Comment by koekeishiya** [comment]
Window borders will be hidden while windows are animating. Attempting to do otherwise will add a ton of complexity and prone to use-after-free issues.

**Comment by FelixKratz** [comment]
The EVENT_HANDLER_SLS_WINDOW_RESIZED functions are called several times during an animation, leading to an unnecessary resize of the border and inflicting some lag with large resolution windows. It would probably be best to guard against a border resize during an animation?

**Comment by koekeishiya** [comment]
> The EVENT_HANDLER_SLS_WINDOW_RESIZED functions are called several times during an animation, leading to an unnecessary resize of the border and inflicting some lag with large resolution windows. It would probably be best to guard against a border resize during an animation?

There should only be one call to that event handler for each *set_window_frame* call that causes a window to change size.
The current system basically does the following:

Figure out which windows are involved in the animation.
Grab a snapshot for each window and prepare inital state.
Hide the real window and its border; show the proxy.
Set the window frame of the real window (this triggers the callback that the border responds to, events are currently queued).
Start the animation in a background thread.
Queued events continue being processed by yabai as normal.

When an animation ends, it hides the proxy and shows the real window, triggering the callback to display the border window again.

**Comment by FelixKratz** [comment]
I have tested my suspicion from above and it seems to be indeed caused by the disabled update: https://github.com/FelixKratz/yabai/commit/a58ce7a083bbf1d5f3cbc774c44397408a41270d on border resize

**Comment by koekeishiya** [comment]
I might have an idea to allow borders to animate; will look into it later.

**Comment by koekeishiya** [comment]
My idea appears to work; borders are animating fine, but needs some cleanup.

**Comment by koekeishiya** [comment]
Trying to animate the border directly causes flicker during the end of transition, so the best solution would simply be to create proxy for the window border too, and pass them through the same kind of system that we do for the windows themselves.

**Comment by koekeishiya** [comment]
I think this is getting close to finished now. I squashed all changes into 2 commits on the `animate` branch. It would be nice if someone else besides me could run this for some time and report unexpected behaviour across the entire feature set of yabai.

**Comment by FelixKratz** [comment]
> I think this is getting close to finished now. I squashed all changes into 2 commits on the `animate` branch. It would be nice if someone else besides me could run this for some time and report unexpected behaviour across the entire feature set of yabai.

There still was some flickering occasionally, related to the reordering of windows.
With the following changes the flickering is eliminated: https://github.com/FelixKratz/yabai/commit/08d7314b9d4051856846e6033f19dd6e4101c1bd

**Comment by FelixKratz** [comment]
An unexpected behavior is:
* When Screen Capture Permission is not given, borders will frequently be ordered *above* the window.

**Comment by wernjie** [comment]
Not sure if these are related to whatever that's mentioned here or if I have any configuration issues:

- A topmost floating window's animation proxy doesn't seem to be topmost, so the proxy might go underneath windows and be entirely invisible while a move animation is in progress. To reproduce: with `yabai -m config window_topmost on` and two bsp windows, call `yabai -m window --toggle float --grid 8:8:1:1:6:6` on one of them.

- It might look better if the animation proxy (for whichever window is in focus _after_ the animation) is on top of the same layer throughout the animation. Example: `yabai -m window --focus stack.last --stack east --focus stack.last` should have the eastern window stay on top while it animates onto the stack.

- Insert feedback borders no longer match the window border properties (like width) specified in config.

On a side note, it'd be nice if we can optionally disable the background blur from the borders if we prefer not to use it or if we aren't making windows transparent.

Thanks for all the work on this!

**Comment by koekeishiya** [comment]
> - A topmost floating window's animation proxy doesn't seem to be topmost, so the proxy might go underneath windows and be entirely invisible while a move animation is in progress. To reproduce: with yabai -m config window_topmost on and two bsp windows, call yabai -m window --toggle float --grid 8:8:1:1:6:6 on one of them.

> - There's (sometimes) a flicker where borders will reappear at the previous location at the end of an animation.

> - Hiding and unhiding an app that has windows across multiple spaces causes animation proxies for all spaces to be created and be visible even though the spaces themselves aren't visible.

> - Unhiding an app that has windows may cause it to appear to flicker between spaces with the app's windows (sometimes indefinitely and I'd need to restart yabai to resolve that).

These should all be fixed in the latest commit.

**Comment by koekeishiya** [comment]
Added the ability to configure border HiDPI mode (on/off) and border blur (on/off). Also updated the docs to specify which command needs SIP to be partially disabled -- should make it easier than having to check the wiki listing.

**Comment by koekeishiya** [comment]
Going to merge into master and do whatever remaining changes are required there.

**Comment by koekeishiya** [comment]
> Insert feedback borders no longer match the window border properties (like width) specified in config.

Fixed in latest commit. Insert feedback borders now use both width and radius setting properly.

**Comment by wernjie** [comment]
- I read that this should be fixed but I'm not sure why I'm still seeing some border flicker here

    ![7](https://user-images.githubusercontent.com/26833058/191661604-27e687d0-90c1-49db-bebe-70c98924a103.gif)

- It also appears that the animations disappear completely for me if i warp a window A towards B in this state (i.e. focus window A, then `yabai -m window --warp east`):
    ```
    +-----+-----+
    |     |     |
    +-----+  B  +
    |  A  |     |
    +-----+-----+
    ```
    which leaves a period of time where all windows are invisible.

I'm currently on the latest commit on the master branch.

**Comment by koekeishiya** [comment]
> It also appears that the animations disappear completely for me if i warp a window A towards B in this state (i.e. focus window A, then yabai -m window --warp east):

I can reproduce this with *auto_balance on*.

> I read that this should be fixed but I'm not sure why I'm still seeing some border flicker here

I cannot reproduce this on either of my machines (older Intel x64 running Big Sur and Apple Silicon running Monterey). Not really sure how that would be happening, as resizing happens while the proxy window is showing instead of the window, and then a transactional swap is performed at the end.

**Comment by wernjie** [comment]
> I cannot reproduce this on either of my machines (older Intel x64 running Big Sur and Apple Silicon running Monterey). Not really sure how that would be happening, as resizing happens while the proxy window is showing instead of the window, and then a transactional swap is performed at the end.

Update: It seems like the flicker I encountered isn't linked to the end of the animation at all. It also seemed like having yabai signals might be worsening/delaying the effect and makes the overall operation slower. 

If I slow down the animation duration further, and remove all `window_moved` and `window_resized` signals I hooked up, the flicker looks something like this:

![8](https://user-images.githubusercontent.com/26833058/191700490-29d85c5b-5f30-4321-a3e8-bdd111e41cd4.gif)


The config I used to reproduce the gif above:
(colours and padding set for visibility)
  ```
  yabai -m config \
      top_padding 4 \
      bottom_padding 4 \
      left_padding 4 \
      right_padding 4 \
      window_gap 4 \
      active_window_border_color 0xffffffff \
      normal_window_border_color 0xff555555 \
      window_border on \
      window_border_width 2 \
      window_animation_duration 1.0 \
      window_topmost on \
      layout bsp
  ```
have two windows in bsp, then run `yabai -m window --toggle float --grid 8:8:1:1:6:6` on one of the windows.

**Comment by koekeishiya** [comment]
Right, ,I understand why it is happening. I'll have to rethink parts of the system because it currently messes with window ordering in a way that causes problems when combining commands, e.g having multiple windows of the same application on space 1, and then doing `yabai -m window --space 3 && yabai -m space --focus 3`, when we *restore* windows at the end of the animation, that window steals focus (when we tell macOS to make it visible on screen). This only happens if there is a window of the same application on that space.

**Comment by koekeishiya** [comment]
@wernjie Do you still have this issue on the latest master? I can't reproduce it with the config you posted.

**Comment by koekeishiya** [comment]
I believe all known reported issues should be resolved now on master. Please let me know if you still encounter issues.

> It might look better if the animation proxy (for whichever window is in focus after the animation) is on top of the same layer throughout the animation. 

I think it would become very complicated to try and catch a focused window event and have that impact the state of a live animation.

**Comment by wernjie** [comment]
> @wernjie Do you still have this issue on the latest master? I can't reproduce it with the config you posted.

It seems like the flicker is much improved on the latest master with that config (~9 out of 10 times it no longer shows up). Though, I can still see a flicker frequently when I have window move/resize signals added to update my status bar. 

For now I can make do with the lack of those signals. Do let me know if there's anything else I can provide to help debug the issue.

**Comment by koekeishiya** [comment]
I was able to catch it in a *single frame*, at the end of an animation, on my screen.

**Comment by wernjie** [comment]
That makes sense. The latest commit has also fixed the flicker for me. Thanks!

**Comment by koekeishiya** [comment]
If no new bugs are detected regarding this issue, I plan to do a release within the weekend.

**Comment by koekeishiya** [comment]
Meh think I'll just gamble; I haven't discovered any bugs so far during my normal usage. If there is a bug, simply setting `window_animation_duration` will make it work like v4.0.4 so should be fine.

**Comment by ubuntudroid** [comment]
That's an awesome feature! Thanks a lot! ❤️ 

However, I also have problems with SIP enabled: the animations are only playing very rarely when swapping or moving windows. Only about every 7th (or even more) command triggers an animation. Not predictably though.

**Comment by koekeishiya** [comment]
> Found a bug with the animations: With SIP enabled (and thus no scripting addition loaded), and window_animation_duration set to a non-zero value, there sometimes is a very wonky animation happening:

So this is happening if you perform some action that would cancel an in-progress animation for a window and start a new animation towards a new state. If you set the animation duration to 1.0 seconds, and do `yabai -m window --swap west` followed by `yabai -m window --swap east` shortly after, you will trigger that behaviour.

What's happening is that the first swap command will trigger the start of an animation, however we *need* the scripting addition to hide the real window while we animate a proxy; performing a transactional switch so that it isn't actually obvious to the user that this is what's happening.  Because the scripting-addition is not loaded, the swap doesn't occur, and there is no animation showing.

When an animation is cancelled because a new one is started, the proxy is updated and swaps with the old proxy. This swap does not rely on the scripting addition, which is why those windows suddenly become visible.

I do believe I stated that this particular config option requires SIP to be disabled. However, I agree that maybe there should be some limitation in place preventing the value to be changed from 0.0 on systems where SIP is enabled.

There is no way to allow this functionality with SIP enabled, unfortunately.

**Comment by ubuntudroid** [comment]
> I do believe I stated that this particular config option requires SIP to be disabled. However, I agree that maybe there should be some limitation in place preventing the value to be changed from 0.0 on systems where SIP is enabled.
>
> There is no way to allow this functionality with SIP enabled, unfortunately.

That makes sense. Think I should have read the docs first - just saw it in the changelog and jumped right into it. 😅 My bad!

**Comment by koekeishiya** [comment]
Added a guard that only allows animation_duration to be set if the appropriate SIP flags are disabled.

**Comment by FelixKratz** [comment]
The animation can get fairly stuttery at times. My testing shows that this is connected to disabling the update on border resize and border redraw. Even though animations use a different SLS connection, they seem to affect each other still.

This patch https://github.com/FelixKratz/yabai/commit/a20e36e017e4e3257b9904625a6b0f080d9e1136 makes the animations (especially noticeable on large displays and short animation durations) dramatically smoother for my configuration. It however introduces redraw flickering since the update has not been disabled on redraw.

**Comment by JakeHL** [comment]
Is it normal for there to be a delay before the animations play? I'm using the latest version on macOS 12.6 apple silicon.

After using `yabai -m window --warp west` there's about a 500ms delay before the windows will actually swap.

**Comment by koekeishiya** [comment]
> Is it normal for there to be a delay before the animations play?

This delay is caused by actually setting the change in size/position using the accessibility API, and there is nothing that can be done to speed up that operation. The accessibility API is janky and that's really it - until Apple either improves the API/performance of it, or implements a different system for third-party software to interact with windows/applications.

**Comment by albert-ying** [comment]
> > can window open and close animations be added?
> 
> It might be possible to create a window opening animation; depends on how much delay there is between the window spawning and yabai getting notified, I guess this would likely be implemented as a fade+scale transition.
> 
> Animating on window close is pretty much impossible. The window is already inaccessible from system APIs by the time we are notified that a window has been destroyed.

What if we let the yabai to send the window closing signal? Say implement a command like `yabai -m close_current_window` and assign `cmd+w` to it using shkd

**Comment by koekeishiya** [comment]
> What if we let the yabai to send the window closing signal? Say implement a command like yabai -m close_current_window and assign cmd+w to it using shkd

That would probably work, but I'd prefer something more universal. I often use `cmd+q` instead of `cmd+w` if it is the last window of an application, and that would not trigger the animation with this solution.

Not that I am fully opposed to experimenting with something like what you mentioned.

**Comment by FelixKratz** [comment]
I have pondered a bit further with the animation system and came up with a new POC for a fade between the scaled window and the unscaled window at the end of an animation. I believe this makes animations a lot more smooth https://github.com/FelixKratz/yabai/commit/6de83963a15f5a2a685f4768b6b3e95f7182bfd0:


https://user-images.githubusercontent.com/22680421/209445475-c867beef-551e-48a3-a2b8-c6af33b8db1f.mp4

**Comment by koekeishiya** [comment]
How does that look when you hit windows that have min/max size constraints?

**Comment by FelixKratz** [comment]
> How does that look when you hit windows that have min/max size constraints?

A bit funny:

https://user-images.githubusercontent.com/22680421/209884625-dedef2b4-9253-40b9-8478-1ef260e5c97e.mp4

**Comment by PrayagS** [comment]
> I have pondered a bit further with the animation system and came up with a new POC for a fade between the scaled window and the unscaled window at the end of an animation. I believe this makes animations a lot more smooth [FelixKratz@de13dbc](https://github.com/FelixKratz/yabai/commit/de13dbca8d62a8a453d15455c7f3adeb527a8e09):

@FelixKratz I'm on the latest release `v5.0.2` and enabled animation but noticed a different behavior on my terminal windows.

The text resizes after the window is set to its new location. In your video, the text resizing and window movement animation happen simultaneously. What could be the reason for this? Does your branch still have a different code?

I'm using the `kitty` terminal. And frame rate is set to 120 in yabai.

https://user-images.githubusercontent.com/44412790/222967164-0023f034-689b-4957-a542-854fe0bffdf2.mp4

**Comment by Lalit64** [comment]
I have the option enabled doesn't seem to be working

**Comment by koekeishiya** [comment]
> Is it normal for there to be a delay before the animations play?
>>. This delay is caused by actually setting the change in size/position using the accessibility API, and there is nothing that can be done to speed up that operation. The accessibility API is janky and that's really it - until Apple either improves the API/performance of it, or implements a different system for third-party software to interact with windows/applications.

This started to annoy me and I got a little bit creative, and found a way to improve the situation.

https://github.com/koekeishiya/yabai/issues/2060

**Comment by net** [comment]
Would love a way to just disable animations for any resize.

**Comment by koekeishiya** [comment]
`window_animation_duration 0.0´ ?

**Comment by net** [comment]
I mean disable only for resize operations, but not for move operations. So any animations that are compromised by the stretching effect, while keeping the animations that look perfect.

**Comment by koekeishiya** [comment]
I'm not sure if that level of complexity can be handled reasonably in the animation system. I agree that the abrupt end from final animated state to displaying the actual window is not ideal.

**Comment by koekeishiya** [comment]
@net 
https://github.com/koekeishiya/yabai/issues/2137

Please try the latest master; is it better or worse in your opinion? Make sure to uninstall the scripting addition and reinstall it after compiling.

**Comment by net** [comment]
I think it's better, but still not something I'd leave enabled. The stretching itself is the issue. I just noticed the stretching causes window corners to deform too when their ratio changes.

The only way I can think to do it, given the constraints, in an aesthetically pleasing way—and this way admittedly would take a lot more work—is to apply a strong gaussian blur to the overlay so you can't really see the window content deforming, and mask the corners so they look like standard macOS corners throughout the full animation.

https://github.com/koekeishiya/yabai/assets/6983821/a9720b29-72c7-465b-9ffb-67ee245a822c

**Comment by net** [comment]
Actually, here's another aesthetically pleasing option, this one even more complicated, but perhaps with the best possible effect:

1. For resizes narrowing a window, do not deform the window image, but rather animate the width down by masking in from the right side.
2. For resizes shortening a window, do the same as above, but masking up from the bottom.
3. For resizes widening a window, take the color of a pixel on the rightmost edge of the window image, preferable near the top just before the corner radius of top right corner begins (to increase the chance of capturing the window's UI background color). Use that color as a background fill color to animate a mask out from the right side of the window image, revealing the background fill as it animates.
4. For resizes lengthening a window, do the same as above, but masking down from the bottom.

A resize may use various combinations of the above options. After the resize, possibly alpha-fade to the window's true UI. The effect of this should be to appear as if the window doesn't relayout its UI until after the animation, but that the window frame itself still cleanly animates.

**Comment by koekeishiya** [comment]
I think that is too much work for the actual processor to do for this to feel snappy, with the way this has to work. Preparing animation state already takes between 60-130ms depending on a variety of conditions.

I think we looked at how Apple performs transitions a bit further up this thread, and they fade transition between the old and new state. Admittedly their version looks better as they have access to more info than yabai does.

**Comment by net** [comment]
Apple does it by performing the alpha fade during the first two thirds of the animation, with some easing function that performs most of the fade within the first few frames. As you noted, this has the effect that by the time the resize animation is slowing down towards the end, the window has already faded to the new state, and so the warping is much less noticeable. They don't even do any corner masking.

I think using the same timings as Apple is as good as anyone can ask for 😄

**Comment by Bellavene** [comment]
> FelixKratz did a more invasive alpha fade POC a couple of comments above. The window transitions approximately in the middle of the animation, so the stretch is less pronounced. Do you think such a solution would be acceptable? I guess it would be hard to tell without trying it yourself and observe the results directly. Looking at a gif is not always the same/as accurate.

Yes! This would be a lot better. Animation is a bit junky right now. Was using fyabai for that reason, until it faded in history.

**Comment by Bellavene** [comment]
Are there any news? Can we bring back without much pain fade that @FelixKratz has made?

**Comment by Melonchanism** [comment]
> Simply calling `SLSSetWindowTransformation` with a `CGAffineTransform` (e..g `CGAffineTransformMakeTranslation(100, 100)` to move a window 100 right and 100 down) will move the window, but this is not animated. Calling this repeatedly will likely cause lag.

[Loop](https://github.com/MrKai77/Loop), another window manager actually uses this (spamming CGAffineTransform). It's not actually that laggy (unless the app you're trying to resize is Music or Weather for some reason), and it could probably be implemented pretty easily. It's even smooth enough to have a little shake effect at the end.

**Comment by Bellavene** [comment]
@Melonchanism 
It is already good enough, it just needs a fade which was already made btw…

**Comment by koekeishiya** [comment]
If you read some of my first comments where I showcase the very first POC I explain why using `SLSSetWindowTransform` is not a valid solution. It is not because of lag.

I problem with fading is that we do not have the final state of the window, only the initial state, so it is not possible to interpolate between the two stages. I attempted to implement the method FelixKratz used, but it suffered from the same limitation I describe in the comments referenced above. A lot of these challengese are mentioned here: https://github.com/koekeishiya/yabai/issues/2137

Maybe if more time and energy was invested it would be possible to figure out a way to do it better than the current yabai version, but I no longer have the interest to keep digging. I've mostly moved on to other personal projects that interest me.

**Comment by Bellavene** [comment]
> Maybe if more time and energy was invested it would be possible to figure out a way to do it better than the current yabai version, but I no longer have the interest to keep digging. I've mostly moved on to other personal projects that interest me.

Yeah, but the animation goes in a fraction of a second, most things you will not even notice. So what @FelixKratz has made was already good enough. There is no need to struggle to make the animation highly perfect, just not like a "frame cut" as it is now. So I would be highly grateful if you just bring back what was already made : )

**Comment by Melonchanism** [comment]
> If you read some of my first comments in this issue where I showcase the very first POC I explain why using `SLSSetWindowTransform` is not a valid solution. It is not because of animation lag.
> 
> The problem with fading is that we do not have the final state of the window, only the initial state, so it is not possible to interpolate between the two stages. I attempted to implement the method FelixKratz used, but it suffered from the same limitation I describe in the comments referenced above. A lot of these challenges are mentioned here: [#2137](https://github.com/koekeishiya/yabai/issues/2137)
> 
> Maybe if more time and energy was invested it would be possible to figure out a way to do it better than the current yabai version, but I no longer have the interest to keep digging. I've mostly moved on to other personal projects that interest me.

They were able to get around it by taking a screenshot, displaying it over everything for 1 frame while the window gets resized to 0, 0 to obtain the minimum size. Then they would move the window back and remove the displayed snapshot and calculate everything.

I just checked now (last time I checked was 1y ago), and they seem to have updated their method. I will check this out.

Another selling point is that this wouldn't require SIP to be disabled (so maybe it could be a separate setting or smth)

**Comment by koekeishiya** [comment]
The 1sec screenshot thing in Loop is to detect a window minimum and maximum size, not for the animation playback itself.

If you want to set the transform of an arbitrary window SIP must be disabled; the macOS API has authorization checks in Sequoia.

The actual problem is that the transform is just that.. a transform; you need to also actually apply the window size changes to the window at some point, it can't just be a transform forever. Actually setting the new frame needs to happen using the AX API which means you need to undo the transform at the same time, failure to do so will also make it look weird.

Removing the transform and adjusting the size using the AX API causes a noticable flicker 1 ~second-ish flicker due to the performance of the AX API, which varies based on which type of content the window is currently rendering. E.g a browser tab playing a youtube video is significantly (up to 500+ ms) slower, which amounts to 31+ frames, which is ... not acceptable..

I have looked at a ton of methods to try and make something better that can work on arbitrary windows, rendering arbitrary content. As I said, it may be possible to do better, but be prepared to invest a large amount of time. Knowing C isn't the bottleeck to this problem, it is interacting with macOS itself that makes this a hard problem.

Make a solution in any programming language of your choice that can animate arbitrary third-party windows on macOS at minimum 60+ fps (Pro-Motion support preferred) and I will port it to C myself, give attribution, and include it in a yabai release.

**Comment by Melonchanism** [comment]
I did a decent bit of testing and found out that spamming AX resize starts to lag after having more than 3-4 animations at once. It seems to only be with resizing and not moving though.

~~My main issue with the current implementation is the problem with window background blur deciding to turn opaque, but it doesn't seem to happen in [FelixKratz' testing version](https://github.com/koekeishiya/yabai/issues/148#issuecomment-1242773105)~~ Ok I read some stuff and its there to prevent flicker from my understanding.

I was looking through some of the dock frameworks, and saw WindowManagement.framework, maybe some functions from this could be used? I know that using the system window manager gives transitions similar to recursively resizing it as Loop does, but they are way smoother and can have several running at the same time.

**Comment by jpatrolla** [comment]
Hi gang,

I’ve been tinkering on my own branch lately and came up with my own POC...

Apologies if I’m covering ground you’ve already explored. I don’t know if I’ve solved anything, but I did at least manage to fix the stretched text issue which was my main goal.. 

What I ended up doing was tweaking /`do_window_scale9()` aka PiP aka `yabai -m window --toggle pip `. You give it  x,y,w,h and it hides the "original" and scales the  window automatically for you according to the points

full disclaimer: I only program as a hobby so please dont judge my code 😅

[do_window_scale_custom](https://github.com/jpatrolla/yabai/blob/89d8c2d7bca9999228f38a777a0465d3203d95e1/src/osax/payload.m#L624)
[animation code](https://github.com/jpatrolla/yabai/blob/89d8c2d7bca9999228f38a777a0465d3203d95e1/src/window_manager.c#L1388)

- On the very first frame :
    - I resize the window  its end size
    - The function creates the pip transform thingamajiggy (toggling on pip)
- On every frame after:
    - The windows position gets updated using `scripting_addition_move_window()` while the animation continues to update `SLSSetWindowTransforn`
    - The nice part is that the PIP hides the window straight off the bat and automatically scales to the bounds, so I (hopefully) I only need to worry about animating/masking the frame/anchoring logic.
    - Since it’s tied directly to the real window, theres no stretching, glitches (yet) and I don’t have to juggle separate proxies. 
-  On the last frame, I “restore” the original transform (which acts like toggling pip off)

PS This is all how I assume it works and It's my first time tinkering with c so please correct me if I've got anything wrong. 
 
The end result: https://github.com/user-attachments/assets/6c504354-c383-43a0-b242-efa998e306ed

Very early days and I’m sure there's room for improvement. Thought I’d share in case it sparks ideas.__

**Comment by jpatrolla** [comment]
Apologies if I’m covering ground anyone's already explored (I’ve read through the posts but I’m not a programmer so Idon’t know if I’ve actually solved anything here.. but thought I’d share in case it was something. 

I tried a couple of ways but what ended up working for me was creating a custom `yabai -m window —toggle pip` function aka `do_window_scale()`  

1.  set_window_frame to the final rects: I do this at the beginning as opposed to the end which addresses the flickering issues caused by move/resize/bsp. If there are any flickers at this point, the animation covers it up well.
2. The pip gets created on frame 0:i use end_w and end_h here so when the last frame calls this command, it thinks its the original and 'restores' the transform (which is where the animation should end up). This allows me to animate the real window in the next step, and saves me from creating proxies and swapping them in/out
- ~~i then update the window’s xy+hw position every frame after with the scripting addition~~ i was initially using scripping addition move/resize to animate the window but realised its enough to just animate via the pip function which sets the  SLSSetWindowTransform. If the window ends up being bigger/smaller it gets scaled automatically. This is where we can get create and actually animate. I still need to tweak some logic for edge cases but for the first iteration the auto/scaling transform works wihtout major issues.
- Since it’s ~~tied directly to~~ the real window, it feels real snappy. Theres no stretching/glitches (yet) and I don’t have to juggle separate proxies
 - On the last frame, the end coordinates get compared in the pip function and “toggle off” it off, restoring the original transform

PS  Sorry if all of this is obvious/self explanatory  - its my first c project I'm just assuming this is how it all works. 
Please correct me if I've got anything wrong.

Lots of ideas to experiment with using this (ie. rebuilding mission control!)

https://github.com/user-attachments/assets/8e9bf432-c7c7-4049-a9a9-0dcf1381c58a

I’m sure there is plenty of room for improvement but I thought I’d share in case it sparks ideas

**Comment by jpatrolla** [comment]
> Make a solution in any programming language of your choice that can animate arbitrary third-party windows on macOS at minimum 60+ fps (Pro-Motion support preferred). Windows that are modified by the same operation should also be animated together as a group (start and end at the same frame). I will port it to C myself, give attribution, and include it in a yabai release.

Does this offer still stand? :D 

https://github.com/user-attachments/assets/f0e9c2d9-62f8-4f7d-8ed0-fdbe9e7fd76a

**Comment by asmvik** [comment]
Post a link to your branch and I'll test and look at it at some point.

With macOS Tahoe releasing this week it will be interesting to see if it works there as well.

Edit: In the gif you posted, I can see the initial resize to the target size, and then it reverts to the original size to play the animation, for a couple of the operations.

**Comment by jpatrolla** [comment]
Still tidying a few things before sharing. I was overthinking parts of it and I’m just now getting a better grasp of the CG/SLS side—it’s given me a new level of respect for the work you’ve done.  I should have something ready to show you in the next day or two

**Comment by Bellavene** [comment]
Any news?

**Comment by jpatrolla** [comment]
Apologies - I got pulled onto another project and probably won't be able to revisit until the end of the month. I made a few steps forward getting an async version working and then took a few steps back implementing `SLSTransactions` before my plate suddenly got full. 
The SLSTransactions version although snappy, brought the stretching right back and I haven't had time to properly experiment since. The code is available on my fork for anyone interested. I think the cleanest version so far is when it was async + regular SLSSetWindowTransform. Will update when I get back to it

**Comment by Explosion-Scratch** [comment]
This is something which I would really love and would really take MacOS to the next level for me. I appreciate all the work jpatrolla is doing on this and just wanted to add my support!

**Comment by jpatrolla** [comment]
@Bellavene didn't expect to be working on it so soon but my yabai broke this morning and while troubleshooting my day turned into [this scene](https://imgur.com/adhd-nutshell-T80xXuA). Long story short I'm now on Tahoe with a fresh yabai install and some clean animations.. but I made a couple of breakthroughts to be fair!.. I present to you: [SLSTransactionSetWindowLockedBounds](https://youtu.be/jg8meamy1nY)

Smooth as butter, no proxies, got rid of my complicated animation/anchor logic(seems like it's baked in) and most importantly there's no stretching. 

@koekeishiya I've finally got my fork to a legible stage: https://github.com/jpatrolla/yabai - so here it is for your review. I'm sure there's still plenty of room for improvement. 
[window_manager](https://github.com/jpatrolla/yabai/blob/3366eefa798237f53aeaf6015a632241510ec40a/src/window_manager.c#L780) -> 
[sa](https://github.com/jpatrolla/yabai/blob/3366eefa798237f53aeaf6015a632241510ec40a/src/sa.m#L545) ->
[payload](https://github.com/jpatrolla/yabai/blob/3366eefa798237f53aeaf6015a632241510ec40a/src/osax/payload.m#L643)

**Comment by Bellavene** [comment]
@jpatrolla 

Looks great! But will it work on sequoia?
And maybe fix that fade to 0.0 in the middle of the animation?

**Comment by jpatrolla** [comment]
> [@jpatrolla](https://github.com/jpatrolla)
> 
> Looks great! But will it work on sequoia? And maybe fix that fade to 0.0 in the middle of the animation?

~I defintely prefer 0.0 myself~ just got what you meant. Yes that's always bothered me when animating fades/opacity. I've baked in a window_animation_min_opacity config for now if people wanted to play with it. I haven't really had a chance to play yet - this is what it look like out of the box. 
It should work as I was using Sequoia when I scraped the list of SLS apis but its untested for now. Its using the latest dev-build so it'll have the latest fixes up til today

**Comment by Bellavene** [comment]
Ah, sorry. Didn't notice in the video, that you changed the parameter of it.

**Comment by Bellavene** [comment]
@jpatrolla 
Tried it on latest sequoia, for some reason scripting additions can not inject into finder, while the original works.

```
could not spawn remote thread: (os/kern) protection failure
yabai: scripting-addition failed to inject payload into Dock.app!
```

Have cloned the code from your repo

**Comment by asmvik** [comment]
> [@Bellavene](https://github.com/Bellavene) didn't expect to be working on it so soon but my yabai broke this morning and while troubleshooting my day turned into [this scene](https://imgur.com/adhd-nutshell-T80xXuA). Long story short I'm now on Tahoe with a fresh yabai install and some clean animations.. but I made a couple of breakthroughts to be fair!.. I present to you: [SLSTransactionSetWindowLockedBounds](https://youtu.be/jg8meamy1nY)
> 
> Smooth as butter, no proxies, got rid of my complicated animation/anchor logic(seems like it's baked in) and most importantly there's no stretching.
> 
> [@koekeishiya](https://github.com/koekeishiya) I've finally got my fork to a legible stage: https://github.com/jpatrolla/yabai - so here it is for your review. I'm sure there's still plenty of room for improvement. [window_manager](https://github.com/jpatrolla/yabai/blob/3366eefa798237f53aeaf6015a632241510ec40a/src/window_manager.c#L780) -> [sa](https://github.com/jpatrolla/yabai/blob/3366eefa798237f53aeaf6015a632241510ec40a/src/sa.m#L545) -> [payload](https://github.com/jpatrolla/yabai/blob/3366eefa798237f53aeaf6015a632241510ec40a/src/osax/payload.m#L643)

I`ll take a look at it at some point, but no ETA, sorry.

**Comment by jpatrolla** [comment]
> I`ll take a look at it at some point, but no ETA, sorry.

No pressure at all (I hope you don’t feel any!)
I’m more than happy to take on the heavy lifting myself — I just need a bit of direction or validation. Even a brief sanity check from someone familiar with C/yabai conventions would help a lot.

It doesn’t have to be from you specifically — anyone who can help ensure the code meets project standards would be great. I’m keen to learn, and even though I’m heavily dependent on AI, I’m very conscious of “AI slop” and do my best to avoid it. Unfortunately, I don’t have access to anyone with direct experience in this space, so I’m mostly self-teaching through experimentation.

Where I’m currently at: the main issue with LockedBounds seems to be that it tends to scale or stretch windows beyond their defined constraints (from SLSWindowIteratorGetConstraints) — ignoring min/max bounds and locked aspect ratios. I’ve started working on mitigation logic using a combo of GetConstraints and Accessibility API fallbacks, which cover most real-world cases but still not 100% bulletproof.

All the private API and reverse-engineering work so far has been a mix of guesswork, testing, and AI-assisted exploration, since there’s so little documentation out there.
I also remember reading [this thread](https://github.com/koekeishiya/yabai/issues/235) a while back — I’ve just revisited it and it looks like there might be some useful information there, so I’ll touch base if I make any updates.

https://github.com/user-attachments/assets/7f320b26-c22f-40eb-b67b-bad5ee9ae495

**Comment by asmvik** [comment]
Note to self:

The scaling issue can actually be fixed with a simple improvement on top of the current system with relative ease.
Do the proxy interpolation as before, and also do the inverse interpolation in reverse while alphablending based on t.

**Comment by SJFCS** [comment]
Could you please let me know the current status of this issue? Is there a compiled build available that we could test or use?

**Comment by jpatrolla** [comment]
I've just revisited my fork afer a long hiatus. I've managed to address the initial flicker and other major issues. It's at a really solid state right now and I'm really happy with how far its come along. Will need to tidy it up before I update my fork but I'll check back here and share it within the next couple of weeks for those interested

**Comment by Explosion-Scratch** [comment]
I'm very interested! I've been watching this issue forever and would absolutely love this - apologies that I don't really have the dev capabilities to contribute to the codebase.

**Comment by jpatrolla** [comment]
@asmvik (re: window_manager_set_window_frame) turns out the AX-performance angle - totally fair read - was a red herring. I think I found the real culprit accidentally while working on the anti-feature of this whole thread 😩🔫 I'll share my findings/solution when I wake up.

**Comment by jpatrolla** [comment]
https://github.com/user-attachments/assets/dbb3d994-006d-4489-8e1a-4c7845a1d712

**Comment by jpatrolla** [comment]
@asmvik please ignore my previous comment/ramblings: they were in regards to a different issue/realisation (see end)...which only surfaced as I was about to push my latest changes. I've just addressed the flow-on and have pushed the latest carve public if you're at all interested: https://github.com/jpatrolla/yabai/

- I've had to move everything payload side in order to keep animations from clashing.
- It's an improvement from my last lockedbounds method: uses a combination of lockedbounds, Transform3D and AX api to achieve animated "true resizes". 
- There's a tiny bit of stretch that rides along due to the AX api's synchronous limitations, but I've managed to get past the biggest hurdles by caching each app's EUI flag, saving a round-trip read per commit. I think it passes for the apps/windows I've tested it against (Chrome, iTerm2, Finder, Safari, Ghostty, kitty) especially for the responsiveness
- Screen recording permissions no longer required since it doesn't utilise any proxy windows
- It's probably more complicated than it needs to be because I've worked other features like focus ring and my 'stage manager' wip into it, but the bones are there and I'd be happy to carve out a specific shape for a PR with some direction if you think I'm capable
- There's also a space ops fix using SLS SPIs in the fork. Not completely drift-proof but potentially an improvement

<details>
<summary>what the earlier ramblings were actually about</summary>

my previous comment was in regards to your ax resize > move > resize sandwich in `window_manager_set_window_frame` — I had assumed the function was structured that way because of the AX api's performance, but the real limitation is AppKit's NSWindow/NSScreen frame clamping preventing the AX api from resizing a window past the screen's bounds. It was a huge facepalm moment for me, and that's when I commented. I'd gotten around it once I realised by moving first and then resizing — one less AX commit than the sandwich — but then I connected my iPad and ran into a second, sneakier variant of the same constraint: display seams. AppKit will also refuse a resize whose landing frame would sit on or near the shared edge between two displays — the AX call reports success but the new size silently never applies, even when the target frame is only a few pixels clear of the seam on one display. Moves are never restricted, only resizes, so the cure is ordering again: resize the window while it's well clear of the seam, then move it into place — which walked me straight back to the resize-before-move ordering your sandwich had all along, just without the second resize commit. Both of these issues have been addressed in the fork.

</details>

**Comment by Melonchanism** [comment]
### Note: Completely unrelated method-wise
I found a fun little bug by using the trackpad to not fully exit mission control in macOS 27. It seems that mission control now continuously updates windows, such as capturing newly created ones in the middle. It creates some decent-looking window animations, although they are not customizable and are still buggy.S

Side note: Dock seems to no longer post mission control accessibility notifications

https://github.com/user-attachments/assets/42a4de7c-5a8f-41ad-a5f6-02ef0f788673

**Comment by jpatrolla** [comment]
An update: I have a method allowing fully native resizing but it requires injection into WindowManager (not AppKit). Allows real resizes meaning any animation code can live daemon side - no AX/SetWindowTransform3D/SetWindowLockedBounds required anywhere.
I've been using it on my private build for about the last month and it's been stable. No issues (that I've noticed). 
 
I haven't made it public but more than happy to share it if you're at all interested @asmvik

**Comment by asmvik** [comment]
> An update: I have a method allowing fully native resizing but it requires injection into WindowManager (not AppKit). Allows real resizes meaning any animation code can live daemon side - no AX/SetWindowTransform3D/SetWindowLockedBounds required anywhere.
> I've been using it on my private build for about the last month and it's been stable. No issues (that I've noticed). 
>  
> I haven't made it public but more than happy to share it if you're at all interested @asmvik

Not sure if I am interested in using it, but it is always nice to share knowledge within this domain. I am sure there are some people who would be interested.

**Comment by Melonchanism** [comment]
> An update: I have a method allowing fully native resizing but it requires injection into WindowManager (not AppKit). Allows real resizes meaning any animation code can live daemon side - no AX/SetWindowTransform3D/SetWindowLockedBounds required anywhere. I've been using it on my private build for about the last month and it's been stable. No issues (that I've noticed).
> 
> I haven't made it public but more than happy to share it if you're at all interested @asmvik

Previously in this issue I speculated about using functions from the WindowManager / WindowManagement frameworks, but never got to any sort of implementation. I did see some of them being imported by Dock so maybe there is some way to do it without injection into WindowManager? Either way I'm interested.
